File: server/AppConfig.ts
import type { EncoderConfig } from "./security/EncoderConfig.js";
import { createRequire } from "node:module";
import { AuthenticationConfiguration } from "./security/AuthenticationConfiguration.js";
import * as fs from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

function isFileAccessible(file: fs.PathLike) {
    try {
        fs.accessSync(file, fs.constants.R_OK);
        return true;
    } catch (e) {
        console.warn(`An error occurred checking if ${file} is accessible.`, e);
        return false;
    }
}

export interface AppConfig {
    authentication: AuthenticationConfiguration;
    db: { file: string };
    security: {
        encoder: EncoderConfig;
        wellKnownLocation: string;
    };
}

let config = {
    security: {
        encoder: {
            saltGenerations: 10,
        },
    },
} as AppConfig;

const appConfigFile = join(__dirname, "./app-config.json");
if (isFileAccessible(appConfigFile)) {
    const require = createRequire(import.meta.url);
    config = Object.assign(config, require(appConfigFile));
}

export default {
    authentication: {
        secret: process.env.APP_AUTHENTICATION_SECRET ?? config.authentication.secret,
    },
    db: {
        file: process.env.SQLITE_DB_FILE ?? config.db.file ?? "pics.db",
    },
    security: {
        encoder: {
            saltGenerations: process.env.SALT_GENERATIONS ?? config.security.encoder.saltGenerations,
        },
        wellKnownLocation: process.env.WELL_KNOWN_LOCATION ?? config.security.wellKnownLocation,
    },
} as AppConfig;

File: server/src/app_config.rs
use serde::Deserialize;

use crate::connection_config::ConnectionConfiguration;
use crate::security::jwt_token_management::AuthenticationConfiguration;
use crate::security::text_encoder::EncoderConfig;

#[derive(Deserialize)]
pub struct SecurityConfiguration {
    pub encoder: EncoderConfig,

    #[serde(default = "default_well_known_location")]
    pub well_known_location: Option<String>,
}

fn default_well_known_location() -> Option<String> {
    None
}

#[derive(Deserialize)]
pub struct AppConfig {
    pub authentication: AuthenticationConfiguration,
    pub db: ConnectionConfiguration,
    pub security: SecurityConfiguration,
}
